src/session/SessionManager.h

```diff
@@ -0,0 +1,130 @@
+/* Copyright (c) 2018 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#ifndef SESSION_SESSIONMANAGER_H_
+#define SESSION_SESSIONMANAGER_H_
+
+#include "common/base/Base.h"
+#include "common/thrift/ThriftTypes.h"
+#include "common/base/StatusOr.h"
+#include "common/thread/GenericWorker.h"
+#include "common/clients/meta/MetaClient.h"
+#include "common/interface/gen-cpp2/GraphService.h"
+
+#include "session/ClientSession.h"
+#include "service/RequestContext.h"
+#include "service/QueryEngine.h"
+
+/**
+ * SessionManager manages the client sessions, e.g. create new, find existing and drop expired.
+ */
+
+namespace nebula {
+namespace graph {
+
+class SessionManager final {
+public:
+    SessionManager(meta::MetaClient* metaClient, const HostAddr &hostAddr);
+    ~SessionManager();
+
+    /**
+     * Create a new session
+     */
+    void createSession(const std::string &userName,
+                       const std::string& clientIp,
+                       folly::Executor* runner,
+                       std::unique_ptr<RequestContext<AuthResponse>> rctx);
+    /**
+     * Remove a session
+     */
+    void removeSession(SessionID id);
+
+    /**
+     * Find a session
+     */
+    void findSession(SessionID id,
+                     folly::Executor* runner,
+                     std::unique_ptr<RequestContext<ExecutionResponse>> rctx,
+                     QueryEngine *queryEngine);
+
+private:
+    using SessionPtr = std::shared_ptr<ClientSession>;
+    /**
+     * Find an existing session
+     */
+    SessionPtr findSessionFromCache(SessionID id);
+
+    using ExecFunc = std::function<void(std::unique_ptr<RequestContext<ExecutionResponse>>)>;
+    void findSessionFromMetad(SessionID id,
+                              folly::Executor* runner,
+                              std::unique_ptr<RequestContext<ExecutionResponse>> rctx,
+                              ExecFunc execFunc);
+
+    void threadFunc();
+
+    void reclaimExpiredSessions();
+
+    void UpdateSessionsToMeta();
+
+
+private:
+    class AuthInstance {
+    public:
+        explicit AuthInstance(std::unique_ptr<RequestContext<AuthResponse>> rctx) {
+            rctx_ = std::move(rctx);
+        }
+
+        ~AuthInstance() = default;
+
+        RequestContext<AuthResponse>* rctx() const {
+            return rctx_.get();
+        }
+
+        void finish() {
+            delete this;
```

Comment:
I think `AuthInstance` and `ExecuteInstance` are over designed. They only wraps a `RequestContext` but nothing else todo. 
```
        void finish() {
            delete this;
        }
```
And this is not a good practice i think, a right value reference is enough for your scenarios. It will easily lead to a double free problem, since the `finish` is a public interface.
```
AuthInstance auth;
auth->finish();
```

Reply:
They are private classes and only managed by the SessionManager, which guarantees them.

Reply:
Did you miss the `over designed` part?